kit/kitsend.man
.\" $Id$
.\"
.\"  Copyright (c) 1991-1997, 2004-2006, Raphael Manfredi
.\"  
.\"  You may redistribute only under the terms of the Artistic Licence,
.\"  as specified in the README file that comes with the distribution.
.\"  You may reuse parts of this distribution only within the terms of
.\"  that same Artistic Licence; a copy of which may be found at the root
.\"  of the source tree for dist 4.0.
.\"
.\" $Log: kitsend.man,v $
.\" Revision 3.0.1.1  1994/05/06  14:00:11  ram
.\" patch23: documented new -V and -h options
.\"
.\" Revision 3.0  1993/08/18  12:04:26  ram
.\" Baseline for dist 3.0 netwide release.
.\"
.TH KITSEND 1 ram
.SH NAME
kitsend \- sends distribution kits
.SH SYNOPSIS
.B kitsend
[
.B \-hV
] [ \fIkits\fR ] \fIrecipients\fR
.SH DESCRIPTION
.I Kitsend
sends distribution kits made by \fImakedist\fR to some recipients specified
on the command line. If you do not specify any kit list, then all the
kits are sent. Otherwise, only the specified kits will be (re)sent.
.PP
A kit list may include ranges, \fI1-10\fR specifying kits 1 through 10,
and \fI5-\fR meaning kits 5 up to the last one. You may also specify kit
numbers by separating them with commas or spaces, and even mix with ranges,
such as: \fI1 3 5-7 9\fR.
.PP
The \fB\-h\fR switch will print out the usage and \fB-V\fR will print
the version number.
.SH AUTHOR
Harlan Stenn <harlan@mumps.pfcs.com>
.SH "SEE ALSO"
makedist(1), kitpost(1).


ChangeLog
Fri Feb 28 17:48:41 MET 1997   Raphael Manfredi <Raphael_Manfredi@grenoble.hp.com>

. Description:

	This huge set of patch results from an integration of the
	perl5 units into the dist-3.0 package. A few improvements
	of mine are also incorporated here, but this patch set should
	mainly be credited to Andy Dougherty and Chip Salzenberg, who
	have both assumed the perl5 Patch Pumpkin with success and brio.

	I won't document most of the changes in detail. Only the obvious
	ones were kept here. You can look closer to the patches themselves
	to see what really changed. Especially the new units, since you
	now have more symbols to check for in your C code...

	Regenerated Configure with current set of units.

	Added Perl Patch Pumpkin credits. Thanks to Andy Dougherty and
	to Chip Salzenberg for their incredible contributions to
	metaconfig.

	Changed my email address and updated copyright dates

	Added support for src.U. This means you can now run Configure
	from a remote directory to separate the build directory and
	the (possibly read-only) source tree.
	Note that some of your private units may require some changes
	due to that...

	Removed support for NO_PROTOTYPE detection on SCO.

	Added cute quoting trick for wild stringify support, which
	allows non-ANSI stringification to occur.

	Replaced .a with $_a all over the place.
	Likewise for .o replaced by $_o.
	You'll have to do that to your own private units.

	Integrated new units from perl5. I will not document most
	of the perl5-related changes here. People interested can
	look at Andy's notes in the perl5 distribution.

	New USE_BSD_GETPGRP to detect the getpgrp() flavour. The old
	USE_BSDPGRP has been obsoleted in favor of USE_BSD_SETPGRP.
	Indeed, setpgrp() and getpgrp() can have distinct flavors,
	so we need two distinct tests for them.

	Useless units dropped: d_group.U and d_passwd.U.

	Mention that <poll.h> can be included if HAS_POLL defined.

	Improved overlapping copy check.

	Added support for Free_t, the type of free().

	Don't use nm with the GNU C library.
	Added support for Linux shared libs.

	Don't prompt them if the void support is high enough for us
	and don't ask for the include path unless they are on a MIPS.

	Added the srcdir support to "configure" now that we have src.U.

	Documented the runnning environment and the src.U unit in
	the metaconfig man page. The $_a and $_o are also mentionned.
	All the generated scripts must now start with a "$startsh"
	to ensure proper shell execution.

	New "create" and "empty" lint directives.

	New -L option for metaxref to match metaconfig and metalint.

	Documents contents of the message sent by patnotify:
	we now let them know the patch priority and description.

	Added a whole bunch of new units, mostly from the perl5 team.

. Files changed:

	* bin/packinit.SH: Fixed one wrong ':' comment in .package.

	* jmake/files/Jmake.rules: Now handles USRINC for dependencies. 
	Smarter about dependencies computation.

	* jmake/files/Jmake.tmpl: Lex path can now be configured.

	* mcon/U/Begin.U: Added Extractall dependency.

	* mcon/U/Config_sh.U, mcon/U/Oldsym.U, mcon/U/Extract.U,
	mcon/U/patchlevel.U, mcon/U/MailAuthor.U, mcon/U/Chk_MANI.U,
	mcon/U/Config_h.U: Added support for src.U.

	* mcon/U/Configdir.U: Have README explicitly mention the package
	name.

	* mcon/U/make.U, mcon/U/d_strtoul.U, mcon/U/Unix.U,
	mcon/U/d_inetaton.U, mcon/U/sitearch.U, mcon/U/d_sanemcmp.U,
	mcon/U/i_locale.U, mcon/U/d_sigaction.U, mcon/U/d_getpgid.U,
	mcon/U/src.U, mcon/U/sitelib.U, mcon/U/d_brokstat.U, mcon/U/man3dir.U,
	, mcon/U/d_sigsetjmp.U, mcon/U/i_sysstat.U, mcon/U/d_strtod.U,
	mcon/U/d_byacc.U, mcon/U/sh.U, mcon/U/Signal.U, mcon/U/d_gnulibc.U,
	mcon/U/man1dir.U, mcon/U/Extractall.U, mcon/U/i_values.U,
	mcon/U/i_sfio.U: Created.

	* mcon/U/Getfile.U: Getfile script now begins with "startsh".

	* mcon/U/Head.U: Make sure we unset CDPATH for shells that support
	this.  Improved Korn shell detection and handling.

	* mcon/U/Instruct.U: Logname / whoami sequence rewritten to use case.

	* mcon/U/Loc.U: Allow users to specify paths on the command line. 
	Will now substitute cp for ln if not supported.

	* mcon/U/Myread.U: Myread script now starts with a "startsh". 
	Miscellaneous fixes.

	* mcon/U/Oldconfig.U: Added support for src.U.  New OSNAME define. 
	Can now sense new OSes.

	* mcon/U/Options.U: Optdef.sh now starts with a "startsh".  Moved
	some code from Head.U.

	* mcon/U/Whoa.U: Whoa script now starts with leading "startsh".

	* mcon/U/afs.U: Can now explicitly tell Configure whether AFS is
	running.

	* mcon/U/archlib.U: Skip existence checks for archlib.

	* mcon/U/archname.U: Changed the way the archname is mangled from
	uname.

	* mcon/U/byteorder.U: No longer ask the user if the test runs ok.

	* mcon/U/ccflags.U: Removed support for NO_PROTOTYPE detection on
	SCO.  New locincpth variable.  Added info on the "additional ld
	flags" question.

	* mcon/U/cf_who.U: New computation method avoiding use of temporary
	file.

	* mcon/U/cpp_stuff.U: Added cute quoting trick for wild stringify
	support.

	* mcon/U/d_bsdjmp.U, mcon/U/d_fd_set.U, mcon/U/models.U,
	mcon/U/d_scannl.U, mcon/U/d_open3.U, mcon/U/d_closedir.U: Added ?F:
	metalint hint.

	* mcon/U/libs.U, mcon/U/libnm.U, mcon/U/libyacc.U, mcon/U/libflex.U,
	mcon/U/libdbm.U, mcon/U/d_socket.U, mcon/U/d_normsig.U,
	mcon/U/d_crypt.U: Replaced .a with $_a all over the place.

	* mcon/U/d_csh.U: Added full_csh to preserve the full path even when
	portable.

	* mcon/U/d_dosuid.U: Moved unit to TOP via a ?Y: layout directive. 
	Tell them /dev/fd is not about floppy disks.

	* mcon/U/d_gconvert.U: Integrated new unit from perl5.

	* mcon/U/d_getpgrp.U: New USE_BSD_GETPGRP to detect the getpgrp()
	flavour.

	* mcon/U/d_group.U, mcon/U/d_passwd.U: Useless unit dropped.

	* mcon/U/d_usendir.U, mcon/U/d_havetlib.U, mcon/U/errnolist.U:
	Replaced .a with $_a all over the place.  Likewise for .o replaced by
	$_o.


kit/kitsend.SH
case $CONFIG in
'')
	if test -f config.sh; then TOP=.;
	elif test -f ../config.sh; then TOP=..;
	elif test -f ../../config.sh; then TOP=../..;
	elif test -f ../../../config.sh; then TOP=../../..;
	elif test -f ../../../../config.sh; then TOP=../../../..;
	else
		echo "Can't find config.sh."; exit 1
	fi
	. $TOP/config.sh
	;;
esac
case "$0" in
*/*) cd `expr X$0 : 'X\(.*\)/'` ;;
esac
echo "Extracting kit/kitsend (with variable substitutions)"
cat >kitsend <<!GROK!THIS!
$startperl
	eval 'exec perl -S \$0 \${1+"\$@"}'
		if \$running_under_some_shell;

# $Id$
#
#  Copyright (c) 1991-1997, 2004-2006, Raphael Manfredi
#  
#  You may redistribute only under the terms of the Artistic Licence,
#  as specified in the README file that comes with the distribution.
#  You may reuse parts of this distribution only within the terms of
#  that same Artistic Licence; a copy of which may be found at the root
#  of the source tree for dist 4.0.
#
# Original Author: Harlan Stenn <harlan@mumps.pfcs.com>
#
# $Log: kitsend.SH,v $
# Revision 3.0.1.2  1994/05/06  13:59:57  ram
# patch23: random code cleanup to follow pat tools style
# patch23: made configurable from dist profile
# patch23: now understands -V and -h options
# patch23: mails now flagged with a bulk precedence
# patch23: added X-Mailer header and now calls mailer via open()
#
# Revision 3.0.1.1  1993/08/19  06:42:15  ram
# patch1: leading config.sh searching was not aborting properly
#
# Revision 3.0  1993/08/18  12:04:25  ram
# Baseline for dist 3.0 netwide release.
#

\$orgname='$orgname';
\$mailer='$mailer';
\$version = '$VERSION';
\$patchlevel = '$PATCHLEVEL';
!GROK!THIS!
cat >>kitsend <<'!NO!SUBS!'

$progname = &profile;		# Read ~/.dist_profile
require 'getopts.pl';
&usage unless &Getopts('hV');

if ($opt_V) {
	print STDERR "$progname $version PL$patchlevel\n";
	exit 0;
} elsif ($opt_h) {
	&usage;
}

$orgname = &tilda_expand($orgname);
chop($orgname = `cat $orgname`) if $orgname =~ m|^/|;

&readpackage;

while ($_ = shift) {
	if (/^(kit)?[1-9][\d,-]*$/) {
		s/^kit//;
		push(@argv, $_);
	}
	else {
		push(@dest, $_);
	}
}
$dest = join(' ',@dest);
&usage unless $dest;

@ARGV = @argv;

if (-f "$package.kit10") {
	@filelist = <$package.kit[0-9][0-9]>;
}
else {
	@filelist = <$package.kit[0-9]>;
}
pop(@filelist) =~ /(\d+)$/ && ($maxnum = $1 + 0);

if ($#ARGV < 0) {
	$argv = "1-$maxnum";
	@ARGV = $argv;
}
$argv = &rangeargs(@ARGV);
@ARGV = split(' ', $argv);

$argv =~ s/ $//;

if ($#ARGV < 0) {
	die "$progname: no kits specified.\n";
} elsif ($#ARGV) {
	print "$progname: sending $package $baserev kits $argv to $dest...\n";
} else {
	print "$progname: sending $package $baserev kit $argv to $dest...\n";
}

fork && exit;

$opt = '-odq' if $mailer =~ /sendmail/;

until ($#ARGV < 0) {
	$kitnum = shift;

	# Provision for broken mailers...
	@dest = split(' ', $dest);
	while (@smalldest = splice(@dest, 0, 50)) {
		$to = join(', ', @smalldest);	# Sensible To: for sendmail
		$smalldest = join(' ', @smalldest);

		open(MAILER, "|$mailer $opt $smalldest") ||
			die "$progname: can't fork $mailer: $!\n";
		print MAILER
"To: $to
Subject: $package $baserev kit #$kitnum
Precedence: bulk
X-Mailer: dist [version $version PL$patchlevel]
Organization: $orgname

[There are $maxnum kits for $package version $baserev.]

";
		$kitnum = "0$kitnum" if $kitnum < 10 && $maxnum >= 10;
		open(FILE,"$package.kit$kitnum") ||
			die "$progname: can't open $package.kit$kitnum: $!\n";
		while (<FILE>) {
			print MAILER;
		}
		close FILE;
		close MAILER;
		warn "$progname: ERROR mailing of $package.kit$kitnum to $dest\n" if $?;
	}
}

sub usage {
	print STDERR <<EOM;
Usage: $progname [-hV] [kits] dest
  -h : print this message and exit
  -V : print version number and exit
EOM
	exit 1;
}

sub rangeargs {
	local($result) = '';
	local($min,$max,$_);
	while ($#_ >= 0) {
		$_ = shift(@_);
		while (/^\s*\d/) {
			s/^\s*(\d+)//;
			$min = $1;
			if (s/^,//) {
				$max = $min;
			}
			elsif (s/^-(\d*)//) {
				$max = $1;
				if ($max == 0 && $maxnum) {
					$max = $maxnum;
				}
				s/^[^,],?//;
			}
			else {
				$max = $min;
			}
			for ($i = $min; $i <= $max; ++$i) {
				$result .= $i . ' ';
			}
		}
	}
	$result;
}

!NO!SUBS!
$grep -v '^;#' ../pl/package.pl >>kitsend
$grep -v '^;#' ../pl/tilde.pl >>kitsend
$grep -v '^;#' ../pl/profile.pl >>kitsend
chmod +x kitsend
$eunicefix kitsend
